utils/apiRequests.ts

```diff
@@ -1,16 +1,30 @@
 import axios from 'axios';
 import { CreateGift, CreateUser, Gift, User } from '~/shared/types';
+import { useQuery } from '@tanstack/react-query';
+import { handleErrorToast } from './handleToasts';
+import { handleError } from './handleError';
 
 const giftsBaseUrl = '/api/gifts';
 
 /**
  *
  * @returns an array that contains all the gifts as objects
  */
-export async function getAllGifts() {
-  // tässä esimerkiksi käytetty sleeppiä, jotta voi kokeilla miten käyttäytyy 3 sekunnin fetchin takia
-  //await sleep(3000);
-  return (await axios.get(giftsBaseUrl)).data as Gift[];
+export function useGetGifts() {
+  return useQuery({
+    queryKey: ['gifts'],
+    queryFn: async () => {
+      try {
+        return (await axios.get(giftsBaseUrl)).data as Gift[];
+      } catch (e) {
+        handleErrorToast(handleError(e));
+        return [];
+      }
+    },
+    refetchOnWindowFocus: false,
+    refetchOnMount: false,
+    retry: false,
+  });
 }
```

Comment:
`useGetGifts` vaihdettu `getAllGifts`in tilalle

`useQuery` vaatii jonkun muun kuin `undefined` returnin. Laitoin nyt, että jos tulee virhe, palautetaan tyhjä array

Reply:
Tässä on nyt se ongelma, että `isError` ei toimi koska käsittelet virheen query funktiossa. Kannattaa [lukea tämä](https://tanstack.com/query/latest/docs/framework/react/guides/query-functions#handling-and-throwing-errors).

Eli virhettä ei kannata ottaa `catch`llä kiinni tässä. On kolme tapaa näyttää toast:

1. Käytä komponentissa `useEffect` hookia joka seuraa `isError` tilaa.
2. Näytä toast hookin `onError` callbackissä:
```ts
  const { data, error, isError } = useQuery('myData', fetchData, {
    onError: (err) => {
      toast.error(`Error: ${err.message}`);
    },
  });
```

3. Näytä toast keskitetysti kaikille hookeille määrittämällä se queryClientiin:
```ts
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      onError: (err) => {
        toast.error(`Error: ${(err as Error).message}`);
      },
    },
    mutations: {
      onError: (err) => {
        toast.error(`Error: ${(err as Error).message}`);
      },
    },
  },
});
```

> Tuleeko mieleen, että onko jokin näistä asetuksista huono / ei tarpeellinen:
> `
refetchOnWindowFocus: false,
refetchOnMount: false,
retry: false,
`

Sanoisin että näiden asetusten oletusarvot ovat hyvät.

Reply:
Kiitos paljon selvennyksestä! 

Tein 1. tavalla toteutuksen, katsotaan onko hyvä :smile: